### PR TITLE
add support for attention policy

### PR DESCRIPTION
**Pull request by Arcturai**

### CODE REVIEW

File: tf/tfprocess.py [hunk 1]
```diff
@@ -1187,9 +1320,15 @@ def construct_net(self, inputs):
...
             h_fc5 = None
 
-        if self.moves_left:
+        # attention weights added as optional output for analysis -- ignored by backend
+        if self.POLICY_HEAD == pb.NetworkFormat.POLICY_ATTENTION:
+            if self.moves_left:
+                outputs = [h_fc1, h_fc3, h_fc5, attn_wts]
+            else:
+                outputs = [h_fc1, h_fc3, attn_wts]
+        elif self.moves_left:
```
Comment (masterkni6) [inline tf/tfprocess.py:1329]: probably dont need these attn_wts at the end

**Comment by masterkni6** [comment]
there's also no newline at end of some files. Maybe running files through yapf will add them

File: tf/tfprocess.py (hunk 1 above)
Comment (Arcturai) [reply tf/tfprocess.py:1329]: The attention weights are useful for analysis, and for an upcoming visualization tool. Exposing them as an output makes this easier. It shouldn't slow down training noticeably and it is not implemented in the backend.

File: tf/make_model.py [hunk 2]
```diff
@@ -33,4 +33,3 @@
 tfp.save_leelaz_weights(leela_path)
 if tfp.swa_enabled:
     tfp.save_swa_weights(swa_path)
-
```
Comment (Tilps) [inline tf/make_model.py:36]: Revert?

File: tf/model_to_net.py [hunk 3]
```diff
@@ -28,4 +28,3 @@
 tfp.save_leelaz_weights(leela_path)
 if tfp.swa_enabled:
     tfp.save_swa_weights(swa_path)
-
```
Comment (Tilps) [inline tf/model_to_net.py:31]: Revert?

File: tf/net.py [hunk 4]
```diff
@@ -289,14 +340,30 @@ def moves_left_to_bp(l, w):
         weights_name = layers[-1]
         pb_name = None
         block = None
+        encoder_block = None
```
Comment (Tilps) [inline tf/net.py:343]: Rename to pol_encoder_block?

File: tf/tfprocess.py [hunk 5]
```diff
@@ -104,6 +117,13 @@ def __init__(self, cfg):
         self.RESIDUAL_BLOCKS = self.cfg['model']['residual_blocks']
         self.SE_ratio = self.cfg['model']['se_ratio']
         self.policy_channels = self.cfg['model'].get('policy_channels', 32)
+        self.policy_embedding_size = self.cfg['model'].get('policy_embedding_size', self.RESIDUAL_FILTERS)
+        self.encoder_layers = self.cfg['model'].get('encoder_layers', 1)
+        self.encoder_heads = self.cfg['model'].get('encoder_heads', 4)
```
Comment (Tilps) [inline tf/tfprocess.py:122]: default of 4 heads seems an interesting choice - if filters is 64 that would mean only 16 per head in the default.

File: tf/tfprocess.py [hunk 6]
```diff
@@ -119,7 +139,7 @@ def __init__(self, cfg):
         # Scale the loss to prevent gradient underflow
         self.loss_scale = 1 if self.model_dtype == tf.float32 else loss_scale
 
-        policy_head = self.cfg['model'].get('policy', 'convolution')
+        policy_head = self.cfg['model'].get('policy', 'attention')
```
Comment (Tilps) [inline tf/tfprocess.py:142]: Changing defaults generally makes people confused...

File: tf/tfprocess.py [hunk 7]
```diff
@@ -104,6 +117,13 @@ def __init__(self, cfg):
         self.RESIDUAL_BLOCKS = self.cfg['model']['residual_blocks']
         self.SE_ratio = self.cfg['model']['se_ratio']
         self.policy_channels = self.cfg['model'].get('policy_channels', 32)
+        self.policy_embedding_size = self.cfg['model'].get('policy_embedding_size', self.RESIDUAL_FILTERS)
+        self.encoder_layers = self.cfg['model'].get('encoder_layers', 1)
```
Comment (Tilps) [inline tf/tfprocess.py:121]: rename to pol_ prefix both variable and the name in the cfg?
Same for encoder_d_model and encoder_dff

File: tf/net.py [hunk 8]
```diff
@@ -333,16 +400,22 @@ def get_weights_v2(self, names):
             if 'renorm' in name:
                 # Renorm variables are not populated.
                 continue
+            if 'headcount' in tf_name:
+                # headcount is set with set_headcount()
+                continue
 
-            pb_name, block = self.tf_name_to_pb_name(name)
+            pb_name, block, encoder_block = self.tf_name_to_pb_name(name)
```
Comment (Tilps) [inline tf/net.py:407]: Similarly rename here and in fill_net_v2.

File: tf/train.py [hunk 9]
```diff
@@ -215,7 +215,7 @@ def main(cmd):
         test_dataset = test_dataset.with_options(options)
         if validation_dataset is not None:
             validation_dataset = validation_dataset.with_options(options)
-    tfprocess.init(train_dataset, test_dataset, validation_dataset)
+    tfprocess.init(train_dataset, test_dataset, validation_dataset)  # None, None, None
```
Comment (Tilps) [inline tf/train.py:218]: Revert?

**Review by Tilps — APPROVED** [review]

**Pull request merged**